File: ProblemSet_20xx/Problem_2054/solution.cs
Solution s = new();
Console.WriteLine(s.MaxTwoEvents([[1,3,2],[4,5,2],[2,4,3]]));
Console.WriteLine(s.MaxTwoEvents([[1,3,2],[4,5,2],[1,5,5]]));
Console.WriteLine(s.MaxTwoEvents([[1,5,3],[1,5,1],[6,6,5]]));
public class Solution {
    public int MaxTwoEvents(int[][] events) {
        Array.Sort(events, (a, b) => a[1] - b[1]);
        int[] maxbeforeindex = new int[events.Length];
        maxbeforeindex[0] = events[0][2];
        int max = events[0][2];
        for (int i = 1; i < events.Length; i++)
        {
            int[] e = events[i];
            int left = 0;
            int right = i - 1;
            int index = -1;
            while (left <= right)
            {
                int mid = (left + right) / 2;
                if (events[mid][1] < e[0])
                {
                    index = mid;
                    left = mid + 1;
                }
                else
                {
                    right = mid - 1;
                }
            }
            if (index >= 0)
            {
                max = Math.Max(max, maxbeforeindex[index] + e[2]);
            }
            else
            {
                max = Math.Max(max, e[2]);
            }
            maxbeforeindex[i] = Math.Max(e[2], maxbeforeindex[i-1]);
        }
        return max;
    }
}

File: ProblemSet_20xx/Problem_2054/solution.py
from typing import List
from bisect import bisect_left


class Solution:
  def maxTwoEvents(self, events: List[List[int]]) -> int:
    events.sort(key = lambda x: x[1])
    dp_values = [0]
    time_frames = []
    result = 0
    
    for start_time, end_time, value in events:
      index = bisect_left(time_frames, start_time)
      prev_value  = dp_values[index]
        
      if (not time_frames or time_frames[-1] != end_time):
        time_frames.append(end_time)
        dp_values.append(max(dp_values[-1], value))
      else:
        dp_values[-1] = max(dp_values[-1], value)
      result = max(result, prev_value + value)
      
    return result

# print(bisect_right([5],4))
s = Solution()
# print("Result", s.maxTwoEvents([[1,3,2],[4,5,2],[2,4,3]]))
# print("Result", s.maxTwoEvents([[1,3,2],[4,5,2],[1,5,5], [7,8,1],[9,10,2]]))
# print("Result", s.maxTwoEvents([[1,5,3],[1,5,1],[6,6,5]]))
print("Result", s.maxTwoEvents([[10,83,53],[63,87,45],[97,100,32],[51,61,16]]))